File: ts-src/core/context.ts
/**
 * LambdaLLM Context - passed to handler functions.
 */

import { ModelConfig, ModelResponse, calculateCost } from './models';
import { ModelInvocationError, TimeoutError } from './exceptions';

export class LambdaLLMContext {
  public model: ModelConfig;
  public timeoutBuffer: number;
  public maxRetries: number;
  public fallbackModel?: ModelConfig;
  private lambdaContext: any;
  private _totalCost = 0;
  private _invocationCount = 0;
  private _provider: any = null;

  constructor(options: {
    model: ModelConfig;
    timeoutBuffer?: number;
    maxRetries?: number;
    fallbackModel?: ModelConfig;
    lambdaContext?: any;
  }) {
    this.model = options.model;
    this.timeoutBuffer = options.timeoutBuffer || 5;
    this.maxRetries = options.maxRetries || 3;
    this.fallbackModel = options.fallbackModel;
    this.lambdaContext = options.lambdaContext;
  }

  get remainingTimeMs(): number {
    if (this.lambdaContext?.getRemainingTimeInMillis) {
      return this.lambdaContext.getRemainingTimeInMillis();
    }
    return 900_000;
  }

  get shouldCheckpoint(): boolean {
    return this.remainingTimeMs < this.timeoutBuffer * 1000;
  }

  get totalCost(): number {
    return this._totalCost;
  }

  get invocationCount(): number {
    return this._invocationCount;
  }

  async invoke(prompt: string, variables?: Record<string, any>): Promise<string> {
    if (this.shouldCheckpoint) {
      throw new TimeoutError('Approaching Lambda timeout');
    }

    const formatted = variables ? this.formatPrompt(prompt, variables) : prompt;
    const response = await this.invokeWithRetry(formatted, this.model);

    this._invocationCount++;
    this._totalCost += response.costUsd;

    return response.content;
  }

  async invokeStructured(prompt: string, variables?: Record<string, any>): Promise<any> {
    const schemaInstruction = '\n\nRespond in valid JSON format only.';
    const formatted = variables ? this.formatPrompt(prompt, variables) : prompt;
    const responseText = await this.invoke(formatted + schemaInstruction);

    try {
      let jsonStr = responseText;
      if (jsonStr.includes('```json')) {
        jsonStr = jsonStr.split('```json')[1].split('```')[0];
      } else if (jsonStr.includes('```')) {
        jsonStr = jsonStr.split('```')[1].split('```')[0];
      }
      return JSON.parse(jsonStr.trim());
    } catch {
      // Retry with explicit instruction
      const retryResponse = await this.invoke(formatted + '\n\nReturn ONLY valid JSON.');
      return JSON.parse(retryResponse.trim());
    }
  }

  getRawClient(service = 'bedrock-runtime'): any {
    // Escape hatch: return raw AWS SDK client
    try {
      const { BedrockRuntimeClient } = require('@aws-sdk/client-bedrock-runtime');
      return new BedrockRuntimeClient({ region: this.model.region || 'us-east-1' });
    } catch {
      throw new Error('Install @aws-sdk/client-bedrock-runtime for raw client access');
    }
  }

  private formatPrompt(template: string, variables: Record<string, any>): string {
    let result = template;
    for (const [key, value] of Object.entries(variables)) {
      result = result.replace(new RegExp(`\\{${key}\\}`, 'g'), String(value));
    }
    return result;
  }

  private async invokeWithRetry(prompt: string, config: ModelConfig): Promise<ModelResponse> {
    let lastError: Error | null = null;

    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        return await this.callProvider(prompt, config);
      } catch (e: any) {
        lastError = e;
        if (e instanceof ModelInvocationError && !e.retryable) throw e;
        const waitMs = Math.min(Math.pow(2, attempt) * 500, 10000);
        await new Promise(resolve => setTimeout(resolve, waitMs));
      }
    }

    if (this.fallbackModel) {
      try {
        return await this.callProvider(prompt, this.fallbackModel);
      } catch { /* fallback also failed */ }
    }

    throw new ModelInvocationError(
      `All ${this.maxRetries} retries exhausted. Last error: ${lastError?.message}`,
      false
    );
  }

  private async callProvider(prompt: string, config: ModelConfig): Promise<ModelResponse> {
    const start = Date.now();

    try {
      const { BedrockRuntimeClient, InvokeModelCommand } = require('@aws-sdk/client-bedrock-runtime');

      if (!this._provider) {
        this._provider = new BedrockRuntimeClient({ region: config.region || 'us-east-1' });
      }

      const body = this.buildRequestBody(prompt, config);
      const command = new InvokeModelCommand({
        modelId: config.modelId,
        body: JSON.stringify(body),
        contentType: 'application/json',
        accept: 'application/json',
      });

      const response = await this._provider.send(command);
      const responseBody = JSON.parse(new TextDecoder().decode(response.body));
      const { content, tokensIn, tokensOut } = this.parseResponse(responseBody, config.modelId);
      const latencyMs = Date.now() - start;
      const costUsd = calculateCost(config.modelId, tokensIn, tokensOut);

      return { content, modelId: config.modelId, tokensIn, tokensOut, latencyMs, costUsd };
    } catch (e: any) {
      if (e.name === 'ThrottlingException') {
        throw new ModelInvocationError(`Bedrock throttled: ${e.message}`, true);
      }
      throw new ModelInvocationError(`Bedrock invocation failed: ${e.message}`, true);
    }
  }

  private buildRequestBody(prompt: string, config: ModelConfig): any {
    const modelId = config.modelId;
    if (modelId.includes('anthropic')) {
      return {
        anthropic_version: 'bedrock-2023-05-31',
        max_tokens: config.maxTokens || 1024,
        temperature: config.temperature || 0.7,
        messages: [{ role: 'user', content: prompt }],
      };
    } else if (modelId.includes('amazon.titan')) {
      return {
        inputText: prompt,
        textGenerationConfig: {
          maxTokenCount: config.maxTokens || 1024,
          temperature: config.temperature || 0.7,
        },
      };
    } else if (modelId.includes('meta.llama')) {
      return { prompt, max_gen_len: config.maxTokens || 1024, temperature: config.temperature || 0.7 };
    }
    return {
      anthropic_version: 'bedrock-2023-05-31',
      max_tokens: config.maxTokens || 1024,
      messages: [{ role: 'user', content: prompt }],
    };
  }

  private parseResponse(body: any, modelId: string): { content: string; tokensIn: number; tokensOut: number } {
    if (modelId.includes('anthropic')) {
      return {
        content: body.content[0].text,
        tokensIn: body.usage?.input_tokens || 0,
        tokensOut: body.usage?.output_tokens || 0,
      };
    } else if (modelId.includes('amazon.titan')) {
      return {
        content: body.results[0].outputText,
        tokensIn: body.inputTextTokenCount || 0,
        tokensOut: body.results?.[0]?.tokenCount || 0,
      };
    } else if (modelId.includes('meta.llama')) {
      return {
        content: body.generation,
        tokensIn: body.prompt_token_count || 0,
        tokensOut: body.generation_token_count || 0,
      };
    }
    return { content: String(body), tokensIn: 0, tokensOut: 0 };
  }
}

File: src/lambdallm/core/context.py
"""LambdaLLM Context - passed to handler functions.

Provides the interface for users to interact with LLM models,
manage state, and access framework utilities within their handler.
"""

import time
import logging
from dataclasses import dataclass, field
from typing import Any, Optional, Union

from lambdallm.core.models import ModelConfig, ModelResponse
from lambdallm.core.exceptions import ModelInvocationError, TimeoutError

logger = logging.getLogger("lambdallm")


@dataclass
class Metrics:
    """Simple metrics collector that emits to CloudWatch."""

    _data: dict = field(default_factory=dict)

    def record(self, name: str, value: float) -> None:
        """Record a metric value."""
        self._data[name] = value

    def get(self, name: str) -> Optional[float]:
        """Get a recorded metric."""
        return self._data.get(name)

    def all(self) -> dict:
        """Get all recorded metrics."""
        return self._data.copy()


class LambdaLLMContext:
    """Context object passed to handler functions.

    Provides:
    - Model invocation (invoke, chat)
    - Session/state access
    - Metrics recording
    - Timeout awareness
    - Raw client access (escape hatch)
    """

    def __init__(
        self,
        model: Optional[ModelConfig],
        timeout_strategy: str,
        timeout_buffer: int,
        max_retries: int,
        fallback_model: Optional[ModelConfig],
        lambda_context: Any,
        middleware: list,
        session_config: Any = None,
        router: Any = None,
    ):
        self.model = model
        self.timeout_strategy = timeout_strategy
        self.timeout_buffer = timeout_buffer
        self.max_retries = max_retries
        self.fallback_model = fallback_model
        self.lambda_context = lambda_context
        self.middleware = middleware
        self.session_config = session_config
        self.router = router
        self.metrics = Metrics()
        self.checkpoint: Optional[dict] = None
        self._provider = None
        self._session = None
        self._invocation_count = 0
        self._total_cost = 0.0

    @property
    def remaining_time_ms(self) -> int:
        """Milliseconds remaining before Lambda timeout."""
        if self.lambda_context and hasattr(self.lambda_context, "get_remaining_time_in_millis"):
            return self.lambda_context.get_remaining_time_in_millis()
        return 900_000  # Default 15 min for local dev

    @property
    def should_checkpoint(self) -> bool:
        """Whether we're approaching timeout and should save progress."""
        return self.remaining_time_ms < (self.timeout_buffer * 1000)

    @property
    def total_cost(self) -> float:
        """Total cost of all model invocations in this request."""
        return self._total_cost

    @property
    def session(self) -> Optional[Any]:
        """Lazy-loaded session state."""
        if self._session is None and self.session_config:
            # Session loading will be implemented in state module
            pass
        return self._session

    def invoke(self, prompt: str, **kwargs) -> str:
        """Invoke the LLM with a prompt string.

        Args:
            prompt: The prompt template string. Use {key} for variable substitution.
            **kwargs: Variables to substitute into the prompt.

        Returns:
            The model's response text.

        Raises:
            ModelInvocationError: If the model call fails after retries.
            TimeoutError: If approaching Lambda timeout.
        """
        # Check timeout
        if self.should_checkpoint:
            raise TimeoutError("Approaching Lambda timeout")

        # Format prompt with variables
        formatted_prompt = prompt.format(**kwargs) if kwargs else prompt

        # Select model (use router if available)
        model_config = self._select_model(formatted_prompt)

        # Invoke with retries
        response = self._invoke_with_retry(formatted_prompt, model_config)

        # Track costs
        self._invocation_count += 1
        self._total_cost += response.cost_usd
        self.metrics.record("model.invocations", self._invocation_count)
        self.metrics.record("model.total_cost_usd", self._total_cost)
        self.metrics.record("model.tokens_in", response.tokens_in)
        self.metrics.record("model.tokens_out", response.tokens_out)
        self.metrics.record("model.latency_ms", response.latency_ms)

        return response.content

    def invoke_structured(self, prompt: str, output_schema: dict, **kwargs) -> dict:
        """Invoke the LLM and parse response into a structured format.

        Args:
            prompt: The prompt template string.
            output_schema: Expected output schema (dict of field_name: type).
            **kwargs: Variables to substitute into the prompt.

        Returns:
            Parsed response as a dictionary matching the schema.
        """
        import json as json_module

        schema_instruction = f"\n\nRespond in JSON format with these fields: {output_schema}"
        full_prompt = prompt.format(**kwargs) + schema_instruction if kwargs else prompt + schema_instruction

        response_text = self.invoke(full_prompt)

        # Try to parse JSON from response
        try:
            # Handle markdown code blocks
            if "```json" in response_text:
                response_text = response_text.split("```json")[1].split("```")[0]
            elif "```" in response_text:
                response_text = response_text.split("```")[1].split("```")[0]

            return json_module.loads(response_text.strip())
        except json_module.JSONDecodeError:
            # Retry once with explicit JSON instruction
            retry_prompt = f"{full_prompt}\n\nIMPORTANT: Return ONLY valid JSON, no other text."
            response_text = self.invoke(retry_prompt)
            return json_module.loads(response_text.strip())

    def get_raw_client(self, service: str = "bedrock-runtime"):
        """Escape hatch: get the raw boto3 client for direct access.

        Use this when the framework's abstraction doesn't fit your needs.
        """
        import boto3

        region = self.model.region if self.model and self.model.region else "us-east-1"
        return boto3.client(service, region_name=region)

    def _select_model(self, prompt: str) -> ModelConfig:
        """Select the best model based on router rules or default."""
        if self.router:
            return self.router.select(prompt, self)
        return self.model

    def _invoke_with_retry(self, prompt: str, model_config: ModelConfig) -> ModelResponse:
        """Invoke model with exponential backoff retry."""
        last_error = None

        for attempt in range(self.max_retries):
            try:
                return self._call_provider(prompt, model_config)
            except ModelInvocationError as e:
                last_error = e
                if not e.retryable:
                    raise
                # Exponential backoff
                wait_time = min(2**attempt * 0.5, 10)
                logger.warning(
                    f"Model invocation failed (attempt {attempt + 1}/{self.max_retries}), "
                    f"retrying in {wait_time}s: {e}"
                )
                time.sleep(wait_time)

        # All retries exhausted - try fallback model
        if self.fallback_model:
            logger.info(f"Falling back to model: {self.fallback_model.model_id}")
            try:
                return self._call_provider(prompt, self.fallback_model)
            except ModelInvocationError:
                pass

        raise ModelInvocationError(
            f"All {self.max_retries} retries exhausted. Last error: {last_error}",
            retryable=False,
        )

    def _call_provider(self, prompt: str, model_config: ModelConfig) -> ModelResponse:
        """Call the model provider. Lazy-loads the provider on first use."""
        if self._provider is None:
            from lambdallm.providers import get_provider

            self._provider = get_provider(model_config)

        return self._provider.invoke(prompt, model_config)
